src/benchmarks.zig

```diff
@@ -1,349 +1,538 @@
-    try stderr.writeAll("\n");
-    try stderr.context.flush();
-
-    var num_errors: usize = 0;
-    inline for (functions, 0..) |def, fcni| {
-        if (fcni > 0)
-            std.debug.print("---\n", .{});
-
-        inline for (args) |arg| {
-            var runtimes: [max_iterations]u64 = undefined;
+    var is_multi_return = try std.ArrayList(bool).initCapacity(allocator, functions.len);
+    defer is_multi_return.deinit();
+
+    inline for (functions) |def| {
+        var fmt_buf: [512]u8 = undefined;
+        const file_name_average = try std.fmt.bufPrint(&fmt_buf, "{s}/{s}.csv", .{ benchmark_name, def.name });
+        const file_average = try results_dir.createFile(file_name_average, .{ .read = true });
+        defer file_average.close();
+        const writer_average = file_average.writer();
+        logger.debug().logf("writing benchmark results to {s}", .{file_name_average});
+
+        var fmt_buf2: [512]u8 = undefined;
+        const file_name_runtimes = try std.fmt.bufPrint(&fmt_buf2, "{s}/{s}_runtimes.csv", .{ benchmark_name, def.name });
+        const file_runtimes = try results_dir.createFile(file_name_runtimes, .{ .read = true });
+        defer file_runtimes.close();
+        const writer_runtimes = file_runtimes.writer();
+
+        inline for (args, 0..) |arg, arg_i| {
+            const arg_name = if (has_args) arg.name else "_";
+            logger.debug().logf("benchmarking arg: {d}/{d}: {s}", .{ arg_i + 1, args.len, arg_name });
+
+            const benchFunction = @field(B, def.name);
+            // NOTE: @TypeOf guarantees no runtime side-effects of argument expressions.
+            // this means the function will *not* be called, this is just computing the return
+            // type.
+            const arguments = blk: {
+                switch (@typeInfo(@TypeOf(benchFunction))) {
+                    .Fn => |info| {
+                        // NOTE: to know if we should pass in the time unit we
+                        // check the input params of the function, so any multi-return
+                        // function NEEDS to have the time unit as the first parameter
+                        if (info.params.len > 0 and info.params[0].type.? == BenchTimeUnit) {
+                            break :blk switch (@TypeOf(arg)) {
+                                void => .{time_unit},
+                                else => .{ time_unit, arg },
+                            };
+                        } else {
+                            // single return type
+                            break :blk switch (@TypeOf(arg)) {
+                                void => .{},
+                                else => .{arg},
+                            };
+                        }
+                    },
+                    else => unreachable,
+                }
+            };
+            const result_type: type = @TypeOf(try @call(.auto, benchFunction, arguments));
```

Comment:
Should be `ResultType`

Reply:
8948a0cb3b14df8d6d5d65356aefa394a1c27ee8